libbeat/autodiscover/providers/jolokia/discovery.go

```diff
@@ -0,0 +1,296 @@
+package jolokia
+
+import (
+	"encoding/json"
+	"net"
+	"reflect"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/common/bus"
+	s "github.com/elastic/beats/libbeat/common/schema"
+	c "github.com/elastic/beats/libbeat/common/schema/mapstriface"
+	"github.com/elastic/beats/libbeat/logp"
+)
+
+// Jolokia Discovery query
+// {
+//   "type": "query"
+// }
+//
+// Example Jolokia Discovery response
+// {
+//   "agent_version": "1.5.0",
+//   "agent_id": "172.18.0.2-7-1322ae88-servlet",
+//   "server_product": "tomcat",
+//   "type": "response",
+//   "server_vendor": "Apache",
+//   "server_version": "7.0.86",
+//   "secured": false,
+//   "url": "http://172.18.0.2:8778/jolokia"
+// }
+//
+// Example discovery probe with socat
+//
+//   echo '{"type":"query"}' | sudo socat STDIO UDP4-DATAGRAM:239.192.48.84:24884,interface=br0 | jq .
+//
+
+// Message contains the information of a Jolokia Discovery message
+var messageSchema = s.Schema{
+	"agent": s.Object{
+		"id":      c.Str("agent_id"),
+		"version": c.Str("agent_version", s.Optional),
+	},
+	"secured": c.Bool("secured", s.Optional),
+	"server": s.Object{
+		"product": c.Str("server_product", s.Optional),
+		"vendor":  c.Str("server_vendor", s.Optional),
+		"version": c.Str("server_version", s.Optional),
+	},
+	"url": c.Str("url"),
+}
+
+// Event is a Jolokia Discovery event
+type Event struct {
+	Type    string
+	Message common.MapStr
+}
+
+// BusEvent converts a Jolokia Discovery event to a autodiscover bus event
+func (e *Event) BusEvent() bus.Event {
+	event := bus.Event{
+		e.Type:    true,
+		"jolokia": e.Message,
+		"meta": common.MapStr{
+			"jolokia": e.Message,
+		},
+	}
+	return event
+}
+
+// Instance is a discovered Jolokia instance, it keeps information of the
+// last probe it replied
+type Instance struct {
+	LastSeen      time.Time
+	LastInterface *InterfaceConfig
+	Message       common.MapStr
+}
+
+// Discovery controls the Jolokia Discovery probes
+type Discovery struct {
+	sync.Mutex
+
+	Interfaces []InterfaceConfig
+
+	instances map[string]*Instance
+
+	events chan Event
+	stop   chan struct{}
+}
+
+// Start starts discovery probes
+func (d *Discovery) Start() {
+	d.instances = make(map[string]*Instance)
+	d.events = make(chan Event)
+	d.stop = make(chan struct{})
+	go d.run()
+}
+
+// Stop stops discovery probes
+func (d *Discovery) Stop() {
+	d.stop <- struct{}{}
+	close(d.events)
+}
+
+// Events returns a channel with the events of started and stopped Jolokia
+// instances discovered
+func (d *Discovery) Events() <-chan Event {
+	return d.events
+}
+
+func (d *Discovery) run() {
+	var cases []reflect.SelectCase
+	for _, i := range d.Interfaces {
+		// Send a probe on each interface when starting
+		d.sendProbe(i)
+
+		// And then one after each interval
+		ticker := time.NewTicker(i.Interval)
+		defer ticker.Stop()
+		cases = append(cases, reflect.SelectCase{
```

Comment:
I'm not particularly a fan of using `reflect`, it makes the code difficult to read, perhaps this could be solved with a goroutine per interface?

Reply:
Better without `reflect`, yes :) I traslated directly the select of the previous implementation to a dynamic select.